pkg/controller/controlplane/valuesprovider.go

```diff
@@ -457,17 +458,94 @@ func (vp *valuesProvider) GetControlPlaneShootCRDsChartValues(
 // GetStorageClassesChartValues returns the values for the shoot storageclasses chart applied by the generic actuator.
 func (vp *valuesProvider) GetStorageClassesChartValues(
 	_ context.Context,
-	_ *extensionsv1alpha1.ControlPlane,
+	controlPlane *extensionsv1alpha1.ControlPlane,
 	cluster *extensionscontroller.Cluster,
 ) (map[string]interface{}, error) {
 	k8sVersionLessThan119, err := version.CompareVersions(cluster.Shoot.Spec.Kubernetes.Version, "<", openstack.CSIMigrationKubernetesVersion)
 	if err != nil {
 		return nil, err
 	}
+	k8sVersionLessThan112, err := version.CompareVersions(cluster.Shoot.Spec.Kubernetes.Version, "<", "1.12")
+	if err != nil {
+		return nil, err
+	}
 
-	return map[string]interface{}{
-		"useLegacyProvisioner": k8sVersionLessThan119,
-	}, nil
+	providerConfig := api.CloudProfileConfig{}
+	if cluster.CloudProfile.Spec.ProviderConfig != nil {
+		if _, _, err := vp.Decoder().Decode(cluster.CloudProfile.Spec.ProviderConfig.Raw, nil, &providerConfig); err != nil {
+			return nil, errors.Wrapf(err, "could not decode providerConfig of controlplane '%s'", kutil.ObjectName(controlPlane))
+		}
+	}
+
+	values := make(map[string]interface{})
+	if providerConfig.StorageClasses != nil && len(providerConfig.StorageClasses) != 0 {
+		allSc := make([]map[string]interface{}, len(providerConfig.StorageClasses))
+		for i, sc := range providerConfig.StorageClasses {
+			allSc[i] = make(map[string]interface{})
+			allSc[i]["name"] = sc.Name
+			if sc.Default != nil && *sc.Default {
+				allSc[i]["default"] = true
+			}
+			if len(sc.Annotations) != 0 {
+				allSc[i]["annotations"] = sc.Annotations
+			}
+			if len(sc.Labels) != 0 {
+				allSc[i]["labels"] = sc.Labels
+			}
+			if len(sc.Parameters) != 0 {
+				allSc[i]["parameters"] = sc.Parameters
+			}
+			if sc.Provisioner != nil && *sc.Provisioner != "" {
+				allSc[i]["provisioner"] = sc.Provisioner
+			} else {
+				allSc[i]["provisioner"] = "cinder.csi.openstack.org"
+			}
+			if sc.ReclaimPolicy != nil && *sc.ReclaimPolicy != "" {
+				allSc[i]["reclaimPolicy"] = sc.ReclaimPolicy
+			}
+			if sc.VolumeBindingMode != nil && *sc.VolumeBindingMode != "" {
+				allSc[i]["volumeBindingMode"] = sc.VolumeBindingMode
+			}
+		}
```

Comment:
```suggestion
		for i, sc := range providerConfig.StorageClasses {
		       var storageClassValues = map[string]interface{}{
		          "name": sc.Name,
		       }
		
			if sc.Default != nil && *sc.Default {
				storageClassValues["default"] = true
			}
			
			if len(sc.Annotations) != 0 {
				storageClassValues["annotations"] = sc.Annotations
			}
			if len(sc.Labels) != 0 {
				storageClassValues["labels"] = sc.Labels
			}
			if len(sc.Parameters) != 0 {
				storageClassValues["parameters"] = sc.Parameters
			}
			
			storageClassValues["provisioner"] = "cinder.csi.openstack.org"
			if sc.Provisioner != nil && *sc.Provisioner != "" {
			    storageClassValues["provisioner"]  = sc.Provisioner
			}
			
			if sc.ReclaimPolicy != nil && *sc.ReclaimPolicy != "" {
				storageClassValues["reclaimPolicy"] = sc.ReclaimPolicy
			}
			
			if sc.VolumeBindingMode != nil && *sc.VolumeBindingMode != "" {
				storageClassValues["volumeBindingMode"] = sc.VolumeBindingMode
			}
			
			allSc[i] = storageClassValues
		}
```

Just an optional structural suggestion as I found it a bit cleaner.